File: java_client/src/main/java/com/cisc468share/net/TcpServer.java
package com.cisc468share.net;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;

public class TcpServer {

    private final ServerSocket serverSocket;
    private final ConnectionHandler handler;

    /** Use a pre-bound ServerSocket — no race condition between port selection and binding. */
    public TcpServer(ServerSocket serverSocket, ConnectionHandler handler) {
        this.serverSocket = serverSocket;
        this.handler = handler;
    }

    public void start() throws IOException {
        System.out.println("TCP server listening on 0.0.0.0:" + serverSocket.getLocalPort());
        while (true) {
            Socket socket = serverSocket.accept();
            new Thread(() -> handler.handle(socket)).start();
        }
    }
}

File: python_client/net/tcp_server.py

import socket
from threading import Thread

class TCPServer:

    def __init__(self, host: str, port: int, handler):
        self.host = host
        self.port = port
        self.handler = handler

    def start(self):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.bind((self.host, self.port))
        s.listen()

        print(f"TCP server listening on {self.host}:{self.port}")

        while True:
            conn, addr = s.accept()
            Thread(target=self.handler, args=(conn, addr), daemon=True).start()
